docs/pattern-matching.md

```diff
@@ -0,0 +1,152 @@
+# Pattern Matching
+
+HVM-Lang offers pattern matching capabilities. You can use pattern matching in defining rules and with the `match` expression.
+
+In example, definitions are a syntax sugar to match expressions:
+```rust
+(Foo 0 false (Cons h1 (Cons h2 t))) = (A h1 h2 t)
+(Foo 0 * *) = B
+(Foo 1+n false *) = n
+(Foo 1+n true *) = 0
+
+Foo = @arg1 @arg2 @arg3 match arg1 arg2 arg3 {
+  (Foo 0 false (Cons h1 (Cons h2 t))): (A h1 h2 t)
+  (Foo 0 * *): B
+  (Foo 1+n false *): n
+  (Foo 1+n true *): 0
+}
+```
+
+Pattern matching on numbers has two forms.
+With the successor pattern it will expect a sequence of numbers up to the `n+var` pattern:
```

Comment:
If the match is of NumSucc(n) type, we need to cover every case up to n, and can't cover a case greater or equal to n with a Num pattern.

Some examples to show what's ok and what's not:
```
match n { 0: ; 2+: ; } // Err, doesn't cover 1
match n { 0: ; 1+: ; 2+: ; } // Err, more than one value of + pattern
match n { 0: ; 1+: ; 1+: ; } // Ok, only one vaue of + pattern
match n { 0: ; 10+: ; n: ; } // Ok, 1 through 9 will go to the default
match n { n: ; 10+: ; } // Ok, all cases will go to 'n', compile to a Var match
match n { 10+: ; n: ; } // Ok, 0 through 9 go to 'n'. Very inneficient.
match n { 0: ; 1: ; 1+: ; } // Err, nums >= 1 can't be covered by a Num pattern
```

We should encourage users to avoid doing strange things with this type of match since it's not very intuitive.